### ISSUE TITLE
How to use it in vue.js project?

### ISSUE THREAD

**Issue by linghuam** [systemjs/systemjs#2235]
- SystemJS Version: 6.5.0
- Which library are you using?
  - [x] system.js
  - [ ] s.js
  - [ ] system-node.cjs
- Which extras are you using?
  - [ ] AMD extra
  - [ ] Named Exports
  - [ ] Named Register
  - [ ] Transform
  - [ ] Use Default
  - [ ] Global
  - [x] Dynamic Import Maps
- Are you using any [custom hooks](/docs/hooks.md)?

### Question
I create a default project use [vue-cli](https://cli.vuejs.org/config/)。I want to use system.js to dynamic import remote packages when click button,so I change index.html & App.vue file. like this:

index.html
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="icon" href="<%= BASE_URL %>favicon.ico">
    <title><%= htmlWebpackPlugin.options.title %></title>
  </head>
  <body>
    <noscript>
      <strong>We're sorry but <%= htmlWebpackPlugin.options.title %> doesn't work properly without JavaScript enabled. Please enable it to continue.</strong>
    </noscript>
    <script type="systemjs-importmap">
      {
        "imports": {
          "vue": "http://127.0.0.1:8081/node_modules/vue/dist/vue.js",
          "lodash": "http://127.0.0.1:8081/node_modules/lodash/lodash.js",
          "com1": "http://127.0.0.1:8080/dist/bundle.js"
        }
      }
    </script>
    <script src="http://127.0.0.1:8081/node_modules/systemjs/dist/system.js"></script>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>

```
App.vue
```js
<template>
  <div id="app">
    <img alt="Vue logo" src="./assets/logo.png">
    <button @click="handleClick">click me to import packages</button>
    <HelloWorld msg="Welcome to Your Vue.js App"/>
  </div>
</template>

<script>
import HelloWorld from './components/HelloWorld.vue'

export default {
  name: 'App',
  components: {
    HelloWorld
  },

  methods: {
    handleClick() {
      System.import('com1').then(res => {
        console.log('import success', res)
      });
    }
  }
}
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
  margin-top: 60px;
}
</style>

```
error output :
```bash
This dependency was not found:

* com1 in ./node_modules/cache-loader/dist/cjs.js??ref--12-0!./node_modules/babel-loader/lib!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader/lib??vue-loader-options!./src/App.vue?vue&type=script&lang=js&

To install it, you can run: npm install --save com1
```
my question is:
1. Is there a problem with my usage？
2. How to use it in vue？

Thank you very much!

**Comment by joeldenning** [comment]
Hi @linghuam, good question. Yes SystemJS can be used with Vue apps. Here are a few things to note.

1. You need to mark all in-browser systemjs modules as [webpack externals](https://webpack.js.org/configuration/externals/#root). In your case, this is `com1`, `lodash`, and `vue`. Marking as external tells webpack to let systemjs provide the module in the browser instead of searching node_modules for it. You can do this in your vue.config.js via the [`configureWebpack`](https://cli.vuejs.org/dev-guide/plugin-api.html#configurewebpack) or [`chainWebpack`](https://cli.vuejs.org/dev-guide/plugin-api.html#chainwebpack) properties. Here is [an example](https://github.com/vue-microfrontends/rate-dogs/blob/97489e2acb1de44aca910ef5e3e0a9d2494200c7/vue.config.js#L14)

2. You should change webpack's [output.libraryTarget](https://webpack.js.org/configuration/output/#outputlibrarytarget) to be [a format supported by SystemJS](https://github.com/systemjs/systemjs/blob/master/docs/module-types.md). Generally the best format to use is `System.register` format, which is supported by setting output.libraryTarget to `"system"`. The `"umd"` libraryTarget also works. See [example](https://github.com/single-spa/vue-cli-plugin-single-spa/blob/21e3ac42f2fad579d3c1864b3b9b3d02e93dc2fe/index.js#L14)

3. You should add the AMD and named-exports SystemJS examples, as they facilitate loading Vue as a systemjs module (since Vue is in UMD format, which the AMD extra supports). See https://github.com/vue-microfrontends/root-config/blob/73a68be26e3a7d27924a1a338a61dd7f1aa487d3/src/index.ejs#L20-L22

4. You should decide whether to inline the script tag for your vue app into the index.html file. Generally with SystemJS you do not put a script tag into the index.html for your module, but instead call `System.import()` and let SystemJS append the script tag. Vue CLI automatically appends the script tags to your index.html as described in the comment at the bottom of the file. You can turn off this behavior by telling [HTMLWebpackPlugin](https://webpack.js.org/plugins/html-webpack-plugin/#root) to not [inject](https://github.com/jantimon/html-webpack-plugin#options) by setting `"inject"` to `false`. If you do let webpack inject the script tag, you'll need to use SystemJS' new "auto import" feature, as described in https://github.com/systemjs/systemjs/releases/tag/6.4.0

**Comment by linghuam** [comment]
Succeeded according to your method,but I found that it was not loaded the way I expected.

Here is a simplified example: [example code](https://github.com/linghuam/system-vue-test)

first,  I modified the index.html file.
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <link rel="icon" href="<%= BASE_URL %>favicon.ico">
    <title><%= htmlWebpackPlugin.options.title %></title>
  </head>
  <body>
    <noscript>
      <strong>We're sorry but <%= htmlWebpackPlugin.options.title %> doesn't work properly without JavaScript enabled. Please enable it to continue.</strong>
    </noscript>
    <script type="systemjs-importmap">
      {
        "imports": {
          "com1": "../bundle/bundle.js"
        }
      }
    </script>
    <script src="https://cdn.jsdelivr.net/npm/systemjs/dist/system.js"></script>
    <div id="app"></div>
    <!-- built files will be auto injected -->
  </body>
</html>

```
this is bundle.js file
```js
System.register([], function(__WEBPACK_DYNAMIC_EXPORT__) {
//........
});
```

second, I modified the vue.config.js.

```js
const path = require('path');

module.exports = {
  configureWebpack: {
    externals: ['com1'],
    output: {
      libraryTarget: 'system'
    }
  }
}
```

third, I modified the App.vue file.

```js
<template>
  <div id="app">
    <button @click="handleClick" style="font-size:20px;">click me to import packages</button>
  </div>
</template>

<script>
export default {
  name: 'App',

  methods: {
    handleClick() {
      // eslint-disable-next-line
      System.import('com1').then(res => {
        console.log('import success', res)
      });
    }
  }
}
</script>
```
then I execute `npm run serve`, in my browser network panel, I found the `bundle.js` file already be loaded.  I expected it to load when the button is clicked, so I can dynamically load the module when I need it.

I am very confused about this phenomenon. Hope to see an example of vue in your warehouse [systemjs-examples](https://github.com/systemjs/systemjs-examples) .

**Comment by joeldenning** [comment]
Hi @linghuam, thanks for providing the sample repo. I have created https://github.com/linghuam/system-vue-test/pull/1 which fixes the issue in your repository.

I omitted one important step from my list above - to turn of webpack's parsing of `System.import()`. This is documented at https://github.com/systemjs/systemjs#compatibility-with-webpack and was a simple addition to the vue.config.js configureWebpack section.

Closing this issue since it seems to be resolved. Feel free to comment further or reopen. Also if you have interest in contributing to the systemjs-examples repo with a Vue example, that would be great. I think it would be most suited for the `starter-kits` folder.
